File: java/src/com/trendmicro/deepsecurity/docs/TenantExamples.java
/* 
 * Copyright 2019 Trend Micro.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *	  http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.trendmicro.deepsecurity.docs;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;

import com.trendmicro.deepsecurity.ApiClient;
import com.trendmicro.deepsecurity.ApiException;
import com.trendmicro.deepsecurity.Configuration;
import com.trendmicro.deepsecurity.api.ApiKeysApi;
import com.trendmicro.deepsecurity.api.ComputersApi;
import com.trendmicro.deepsecurity.api.PoliciesApi;
import com.trendmicro.deepsecurity.api.TenantsApi;
import com.trendmicro.deepsecurity.auth.ApiKeyAuth;
import com.trendmicro.deepsecurity.model.Administrator;
import com.trendmicro.deepsecurity.model.ApiKey;
import com.trendmicro.deepsecurity.model.Computer;
import com.trendmicro.deepsecurity.model.Computers;
import com.trendmicro.deepsecurity.model.IntrusionPreventionComputerExtension;
import com.trendmicro.deepsecurity.model.Policy;
import com.trendmicro.deepsecurity.model.SearchCriteria;
import com.trendmicro.deepsecurity.model.SearchFilter;
import com.trendmicro.deepsecurity.model.Tenant;
import com.trendmicro.deepsecurity.model.Tenant.ModulesVisibleEnum;
import com.trendmicro.deepsecurity.model.Tenants;

/**
 * Creates and configures tenants.
 */
public class TenantExamples {

	/**
	 * Creates a tenant on the primary Deep Security Manager
	 * 
	 * @param accountName The name of the tenant account.
	 * @param apiVersion The version of the API to use.
	 * @throws ApiException if a problem occurs when creating the tenant on Deep Security Manager.
	 * @return The new tenant.
	 */
	public static Tenant createTenant(String accountName, String apiVersion) throws ApiException {

		// Create and configure a Tenant object
		Tenant tenant = new Tenant();

		// Set module visibility
		List<ModulesVisibleEnum> modules = new ArrayList<>();
		modules.add(ModulesVisibleEnum.ANTI_MALWARE);
		modules.add(ModulesVisibleEnum.FIREWALL);
		modules.add(ModulesVisibleEnum.INTRUSION_PREVENTION);

		// Administrator account
		Administrator admin = new Administrator();
		admin.setUsername("MasterAdmin");
		admin.setPassword("P@55word");
		admin.setEmailAddress("bad@email.com");

		tenant.setName(accountName);
		tenant.setLocale(Tenant.LocaleEnum.EN_US);
		tenant.setDescription("Test tenant.");
		tenant.setAdministrator(admin);
		tenant.setModulesVisible(modules);

		// Add the tenant to the manager
		TenantsApi tenantsApi = new TenantsApi();
		return tenantsApi.createTenant(tenant, Boolean.TRUE, Boolean.FALSE, Boolean.TRUE, apiVersion);
	}

	/**
	 * Retrieves the state of the Intrusion Prevention module for a tenant's computers
	 * 
	 * @param tenantID The ID of the tenant.
	 * @param roleID The ID of the role to use for the tenant's API key
	 * @param apiVersion The version of the API to use.
	 * @throws ApiException if a problem occurs when deleting and creating API keys, and if a problem occurs when obtaining the list of computers.
	 * @return A Map that uses the computer ID as the key and the intrusion prevention state as the value.
	 */
	public static Map<Integer, IntrusionPreventionComputerExtension.StateEnum> getIPStatesForTenant(Integer tenantID, Integer roleID, String apiVersion) throws ApiException {

		Map<Integer, IntrusionPreventionComputerExtension.StateEnum> computerIPStates = new HashMap<>();

		// Create the key object
		ApiKey key = new ApiKey();
		key.setKeyName("Temporary Key");
		key.setRoleID(roleID);
		key.setLocale(ApiKey.LocaleEnum.EN_US);
		key.setTimeZone("Asia/Tokyo");

		// Add the key to Deep Security Manager
		TenantsApi tenantsApi = new TenantsApi();
		key = tenantsApi.generateTenantApiSecretKey(tenantID, key, apiVersion);

		// Configure an APIClient using the new key's secret
		ApiClient tenantClient = Configuration.getDefaultApiClient();
		ApiKeyAuth DefaultAuthentication = (ApiKeyAuth)tenantClient.getAuthentication("DefaultAuthentication");
		DefaultAuthentication.setApiKey(key.getSecretKey());

		// Get the computers and find the states
		ComputersApi tnComputerApi = new ComputersApi(tenantClient);
		Computers computers;
		computers = tnComputerApi.listComputers(Boolean.FALSE, apiVersion);

		for (Computer computer : computers.getComputers()) {
			computerIPStates.put(computer.getID(), computer.getIntrusionPrevention().getState());
		}

		// Delete the tenant key
		ApiKeysApi tnApiKeysApi = new ApiKeysApi(tenantClient);
		tnApiKeysApi.deleteApiKey(key.getID(), apiVersion);

		return computerIPStates;
	}

	/**
	 * Retrieves the Intrusion Prevention rules that are applied to all computers of all tenants.
	 * 
	 * @param apiVersion The version of the API to use.
	 * @throws ApiException if a problem occurs when searching, if a problem occurs when deleting and creating API keys, and if a problem occurs when listing computers.
	 * @return A Map object that uses the tenant ID as the key, and a Map of rules that are applied to the computers.
	 */
	public static Map<Integer, Map<Integer, ArrayList<Integer>>> getTenantRules(String apiVersion) throws ApiException, IOException {

		// Key is tenant ID. Value is a list of computer rule IDs
		Map<Integer, Map<Integer, ArrayList<Integer>>> tenantMap = new HashMap<>();

		// Key is computer ID. Value is a list of rule IDs
		Map<Integer, ArrayList<Integer>> computerRules = new HashMap<>();

		// Obtain connection properties from local properties file
		Properties properties = new Properties();
		ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
		try (InputStream input = classLoader.getResourceAsStream("com/trendmicro/deepsecurity/docs/Resources/example.properties")) {
			properties.load(input);

			String primarySecretKey = properties.getProperty("secretkey");
			String primaryURL = properties.getProperty("url");

			// Configure the ApiClient
			ApiClient apiClient = Configuration.getDefaultApiClient();
			apiClient.setBasePath(primaryURL);
			ApiKeyAuth defaultAuthentication = (ApiKeyAuth)apiClient.getAuthentication("DefaultAuthentication");
			defaultAuthentication.setApiKey(primarySecretKey);

			// Search for Active tenants
			SearchCriteria searchCriteria = new SearchCriteria();
			searchCriteria.setFieldName("tenantState");
			searchCriteria.setChoiceValue("active");
			searchCriteria.setChoiceTest(SearchCriteria.ChoiceTestEnum.EQUAL);

			// Search filter
			SearchFilter searchFilter = new SearchFilter();
			searchFilter.setMaxItems(Integer.valueOf(1));
			searchFilter.addSearchCriteriaItem(searchCriteria);

			TenantsApi tenantsApi = new TenantsApi();
			Tenants tenants = tenantsApi.searchTenants(searchFilter, apiVersion);

			// Iterate the tenants
			for (Tenant tenant : tenants.getTenants()) {
				// For each tenant create an api key
				ApiKey tenantKey = new ApiKey();
				tenantKey.setKeyName("Temporary Key");
				tenantKey.setRoleID(Integer.valueOf(1));
				tenantKey.setLocale(ApiKey.LocaleEnum.EN_US);
				tenantKey.setTimeZone("Asia/Tokyo");
				// Add the key to Deep Security Manager
				tenantKey = tenantsApi.generateTenantApiSecretKey(tenant.getID(), tenantKey, apiVersion);

				// Configure the ApiClient to use the tenant's secret key
				defaultAuthentication.setApiKey(tenantKey.getSecretKey());

				// Create a ComputersApi object for the tenant
				ComputersApi tnComputersApi = new ComputersApi();

				// Iterate over the tenant computers
				Computers tenantComputers = tnComputersApi.listComputers(Boolean.FALSE, apiVersion);
				for (Computer tenantComputer : tenantComputers.getComputers()) {
					IntrusionPreventionComputerExtension intrusionPeventionComputerExtension = tenantComputer.getIntrusionPrevention();
					computerRules.put(tenantComputer.getID(), (ArrayList<Integer>)intrusionPeventionComputerExtension.getRuleIDs());
				}
				tenantMap.put(tenant.getID(), computerRules);

				// Delete the tenant key
				ApiKeysApi tnApiKeysApi = new ApiKeysApi();
				tnApiKeysApi.deleteApiKey(tenantKey.getID(), apiVersion);
				tenantKey = null;

				// Configure the ApiClient to use the primary tenant's Secret Key
				defaultAuthentication.setApiKey(primarySecretKey);
			}
			return tenantMap;
		}
	}

	/**
	 * Adds a policy to a tenant
	 * 
	 * @param policy The Policy to add to the tenant.
	 * @param tenantID The ID of the tenant.
	 * @param apiVersion The version of the API to use.
	 * @throws ApiException if a problem occurs when deleting and creating API keys and if a problem occurs when creating the policy on Deep Security Manager.
	 * @return The created policy.
	 */
	public static Policy addTenantPolicy(Policy policy, Integer tenantID, String apiVersion) throws ApiException {

		// Create an API key for the tenant
		TenantsApi tenantsApi = new TenantsApi();
		ApiKey tenantKey = new ApiKey();
		tenantKey.setKeyName("Tenant Key");
		tenantKey.setRoleID(Integer.valueOf(1));
		tenantKey = tenantsApi.generateTenantApiSecretKey(tenantID, tenantKey, apiVersion);

		// Create an ApiClient object for the tenant
		ApiClient tenantClient = Configuration.getDefaultApiClient();
		tenantClient.setBasePath("https://localhost:4119/api");
		ApiKeyAuth defaultAuthentication = (ApiKeyAuth)tenantClient.getAuthentication("DefaultAuthentication");
		defaultAuthentication.setApiKey(tenantKey.getSecretKey());

		// Add the policy
		PoliciesApi tnPoliciesApi = new PoliciesApi(tenantClient);
		policy = tnPoliciesApi.createPolicy(policy, Boolean.FALSE, apiVersion);

		// Delete the tenant key
		ApiKeysApi tnApiKeysApi = new ApiKeysApi(tenantClient);
		tnApiKeysApi.deleteApiKey(tenantKey.getID(), apiVersion);

		return policy;
	}
}

File: python/src/tenant_examples.py
# Copyright 2019 Trend Micro.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#


def create_tenant(api, configuration, api_version, api_exception, account_name):
    """ Creates a tenant on the primary Deep Security Manager.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :param account_name: The account name to use for the tenant.
    :return: A TenantsApi object that contains the new tenant.
    """

    # Define the administrator account
    admin = api.Administrator()
    admin.username = "TenantAdmin"
    admin.password = "Pas$w0rd"
    admin.email_address = "example@email.com"
    admin.receive_notifications = "false"
    admin.role_id = 1

    # Create a tenant
    tenant = api.Tenant(administrator=admin)

    # Set the visible modules
    modules = api.Tenant.modules_visible = ["anti-malware", "firewall", "intrusion-prevention"]
    tenant.modules_visible = modules

    # Set the account name
    tenant.name = account_name

    # Set the locale and description
    tenant.locale = "en-US"
    tenant.description = "Test tenant."

    # Create the tenant on Deep Security Manager
    try:
        tenants_api = api.TenantsApi(api.ApiClient(configuration))
        return tenants_api.create_tenant(tenant, api_version, confirmation_required=False, asynchronous=True)

    except api_exception as e:
        return "Exception: " + str(e)


def get_ip_states_for_tenant(api, configuration, api_version, api_exception, tenant_id):
    """ Obtains the running state of the Intrusion Prevention module for a tenant's computers.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :param tenant_id: The ID of the tenant.
    :return: A dictionary that contains computer IDs and the module running state.
    """

    computer_ip_states = {}

    primary_key = configuration.api_key['api-secret-key']

    # Create an API key
    key = api.ApiKey()
    key.key_name = "Temporary API Key"
    key.role_id = 1
    key.locale = "en-US"
    key.time_zone = "Asia/Tokyo"

    try:
        # Check that the tenant is in the 'active' state
        state = api.TenantsApi(api.ApiClient(configuration)).describe_tenant(tenant_id, api_version).tenant_state
        if state == 'active':

            # Generate the secret key for the tenant
            tenants_api = api.TenantsApi(api.ApiClient(configuration))
            generated_key = tenants_api.generate_tenant_api_secret_key(tenant_id, key, api_version)

            # Add the secret key to the configuration
            configuration.api_key['api-secret-key'] = generated_key.secret_key

            # Include Intrusion Prevention information in the returned Computer objects
            expand = api.Expand(api.Expand.intrusion_prevention)

            # Get a list of tenant computers
            computers_api = api.ComputersApi(api.ApiClient(configuration))
            computers_list = computers_api.list_computers(api_version, expand=expand.list(), overrides=False)

            # Find the Intrusion Prevention state for each computer
            for computer in computers_list.computers:
                computer_ip_states[computer.id] = computer.intrusion_prevention.state

            # Reset the API key to the primary key
            configuration.api_key['api-secret-key'] = primary_key

        return computer_ip_states

    except api_exception as e:
        return "Exception: " + str(e)


def get_ip_rules_for_tenant_computers(api, configuration, api_version, api_exception):
    """ Obtains the IDs of the Intrusion Prevention rules that are assigned to each tenant's computers.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :return: A dictionary of tenants IDs that contains a dictionary of computer IDs with the IP rules they are using.
    """

    tenant_rules = {}
    primary_key = configuration.api_key['api-secret-key']

    try:
        tenants_api = api.TenantsApi(api.ApiClient(configuration))
        tenants_list = tenants_api.list_tenants(api_version)

        for tenant in tenants_list.tenants:
            print("Processing tenant " + str(tenant.id))

            #  Check that the tenant is in the 'active' state
            state = api.TenantsApi(api.ApiClient(configuration)).describe_tenant(tenant.id, api_version).tenant_state
            if state == 'active':

                # Create an API key
                key = api.ApiKey()
                key.key_name = "Temporary Key for getting IP rules from tenant computers"
                key.role_id = 1
                key.locale = "en-US"
                key.time_zone = "Asia/Tokyo"

                # Generate the secret key for the tenant
                tenants_api = api.TenantsApi(api.ApiClient(configuration))
                generated_key = tenants_api.generate_tenant_api_secret_key(tenant.id, key, api_version)

                # Add the secret key to the configuration
                configuration.api_key['api-secret-key'] = generated_key.secret_key

                # Include Intrusion Prevention information in the retrieved Computer objects
                expand = api.Expand(api.Expand.intrusion_prevention)

                # Create a ComputersApi object for the tenant
                computers_api = api.ComputersApi(api.ApiClient(configuration))

                # Get a list of computers for the tenant
                computers_list = computers_api.list_computers(api_version, expand=expand.list(), overrides=False)

                # For the tenant, get the IP rules for all computers
                computer_ip_rules = {}
                for computer in computers_list.computers:
                    computer_ip_rules[computer.id] = computer.intrusion_prevention.rule_ids

                tenant_rules[tenant.id] = computer_ip_rules

                # Reset the API key to the primary key
                configuration.api_key['api-secret-key'] = primary_key

        return tenant_rules

    except api_exception as e:
        return "Exception: " + str(e)


def add_policy_to_tenant(api, configuration, api_version, api_exception, policy, tenant_id):
    """ Adds a policy to a tenant.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :param policy: The policy to add to the tenant.
    :param tenant_id: The ID of the tenant.
    :return: A PoliciesApi object that contains the new tenant.
    """

    tenant_client_with_policy = None
    primary_key = configuration.api_key['api-secret-key']

    # Create an API key
    key = api.ApiKey()
    key.key_name = "Temporary key for adding policy to a tenant"
    key.role_id = 1
    key.locale = "en-US"
    key.timeZone = "Asia/Tokyo"

    try:
        # Check that the tenant is in the 'active' state
        state = api.TenantsApi(api.ApiClient(configuration)).describe_tenant(tenant_id, api_version).tenant_state
        if state == 'active':

            # Generate the secret key for the tenant
            tenants_api = api.TenantsApi(api.ApiClient(configuration))
            generated_key = tenants_api.generate_tenant_api_secret_key(tenant_id, key, api_version)

            # Add the secret key to the configuration
            configuration.api_key['api-secret-key'] = generated_key.secret_key

            # Add the policy
            tenant_policies_api = api.PoliciesApi(api.ApiClient(configuration))
            tenant_client_with_policy = tenant_policies_api.create_policy(policy, api_version, overrides=False)

            # Reset the API key to the primary key
            configuration.api_key['api-secret-key'] = primary_key

        return tenant_client_with_policy

    except api_exception as e:
        return "Exception: " + str(e)
